### PR TITLE
Interactive column sizing

### PR DESCRIPTION
**Pull request by atmgrifter00**
# Pull Request

## 🤨 Rationale

- #846

## 👩‍💻 Implementation

Bulk of implementation is the creation of the new `TableLayoutManager` class that is responsible for updating the various aspects of sizing state the table needs.

Beyond that it was mainly just updating the `Table` template to host the new interactive dividers, and a "glass pane" `div` to handle the disabling of hover states in template elements during a column resize operation.

## 🧪 Testing

Added unit tests.

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/table/models/table-layout-manager.ts:240]: You will notice some obvious similarities between this method and `performCascadeSizeRight`. I made some attempts to refactor the methods to share some of the similar code, but didn't come up with something I liked. It's difficult because a lot of it is "reversed" logic. I still believe there is room to at least make some of the logic more consistent between the two, and am very interested in suggestions here.

File: packages/nimble-components/src/table/styles.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/table/styles.ts:118]: These CSS selectors were my admittedly goofy way of dealing with not marking dividers on a column other than the one you started the drag from as `active` (and thus preventing them from appearing). I've already discussed with @mollykreis that whatever solution we come up with from preventing row hover state as you drag a divider should be applicable to other column headers, so my hope is that these CSS selectors (and ultimately the `not-active` attribute) will ultimately go away.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/models/table-layout-manager.ts:99]: Why do you need the `columnIndex` here? I would think knowing which divider was clicked would be enough. For example, what's the difference between clicking on the right divider of column 0 vs the left divider of column 1 (i.e. the difference of passing `0, 0` and `1, 0`)?

File: packages/nimble-components/src/table/index.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/index.ts:562]: Why does the table viewport min width need to be updated when the columns change? Looking at the implementation of `updateTableViewportMinWidth`, I'm guessing it has to do with the calculation for the total minimum column width. Is that correct? If so, wouldn't that code need to be run again in the `update` code path due to a column configuration changing?

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/models/table-layout-manager.ts:174]: Doesn't look like this handles hidden columns.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/models/table-layout-manager.ts:188]: I think this can be simplified
```suggestion
            totalColumnFixedWidth += column.columnInternals.currentPixelWidth ?? 0;
```

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/models/table-layout-manager.ts:207]: I don't think this code handles hidden columns correctly. We don't create the `.header-container` element for hidden columns which means that your `headers` variable isn't guaranteed to be a parallel array to `this.table.columns`.

File: packages/nimble-components/src/table/template.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/table/template.ts:45]: This `div` is responsible for preventing hover states on elements during a column sizing operation. I'm not sure of a cleaner way to provide this behavior. Additionally, this proved difficult to write a unit test for, as I believe we ultimately need an API that actually moves a mouse to a position on screen.

Open to a class name other than `glass-pane` as well.

File: packages/nimble-components/src/table/template.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/table/template.ts:50]: I initially added this wrapper element (and provided it a `ref`) because at one point my implementation needed to more easily track when the size of the combination of the contained elements changed. While this is no longer the case, its presence does clean up a pageObject implementation, so I'm not sure if that is enough to keep this around.

File: packages/nimble-components/src/table/styles.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/table/styles.ts:109]: This CSS allows us to provide a larger hit-test region for the divider without adjusting its visual position. I've started with a width of 8px, which seems pretty forgiving, but am open to suggestion here.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts:32]: This idea actually doesn't work for a grid contained within a flex box. The usage of `minmax` will allow for the grid to be sized smaller than what is specified by the max parameter. We need the element with the `grid` display to size to the full width specified by the max parameter.

File: packages/nimble-components/src/table/index.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/index.ts]: This has been reworked such that we are no longer using a width factor.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts]: There is no longer the need for a `connectedCallback` implementation.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts]: I've reworked the logic to remove this.

File: packages/nimble-components/src/table/index.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/index.ts]: No longer doing this.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts]: We now are handling hidden columns.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts]: Handling hidden columns now.

File: packages/nimble-components/src/table/styles.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/styles.ts]: Removed.

File: change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json
Comment (jattasNI) [inline change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json:1]: One interesting behavior I noticed in initial Storybook testing: if you start a column resize and then release your mouse outside the browser, the drag operation isn't canceled. The result is that if you then bring your mouse back into the browser, the drag operation is still happening even though the mouse is no longer down. I'm seeing this on Safari and Edge on macOS but interestingly not Firefox (unless I'm doing something weird, I was testing in a rush).

Comment (jattasNI) [reply change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json:1]: A similar weird behavior: start the drag with a right click instead of a left click. The browser context menu appears but if you dismiss it there is still a drag operation in progress even though the mouse isn't pressed.

File: change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json
Comment (jattasNI) [inline change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json:2]: I think it'd be worthwhile to add review and update our Storybook docs on table column sizing. I don't see much to change but it might be helpful to modify the parameters for "Fractional width configuration" and `min-pixel-width` to mention that:
1. fractional width is just about initial size and size when the table is resized. It's not updated when you resize a column
2. min-pixel-width is applied when the table resizes or when a column is resized

File: packages/nimble-components/src/table/styles.ts
Comment (jattasNI) [inline packages/nimble-components/src/table/styles.ts:114]: Where does `-5px` come from? Is it possible to express it as a calculation of other named values?

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (jattasNI) [inline packages/nimble-components/src/table/models/table-layout-manager.ts:9]: I didn't look at this file in any amount of detail. I think someone besides you should understand it and take a closer look but it needn't be every reviewer. I think I heard you mention that person was likely to be @mollykreis but I'm happy to step in and get an overview if that isn't true.

File: change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json
Comment (atmgrifter00) [reply change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json:1]: Regarding your first issue, I think this is a Storybook quirk. You don't see the same behavior in the Angular application. Additionally, you don't see the noted behavior in Chrome with vanilla JS apps in StackBlitz that have a similar setup. Not sure how we want to proceed given that.

For the second issue, I think I can just be more guarded in my event handling where I only process left clicks (or the "primary" button).

File: packages/nimble-components/src/table/testing/table.pageobject.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/testing/table.pageobject.ts:482]: Would it be better to write this to return `HTMLElement | null` so that you only throw if you are trying to index past the number of columns, not the number of dividers. Then you can write tests that the right-most column doesn't have a divider on its right -- which currently is a bug in the code based on where the hidden columns fall within the DOM.

File: packages/nimble-components/src/table/tests/table-column-sizing.spec.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/tests/table-column-sizing.spec.ts:193]: I'm guessing you aren't planning on checking this code in with only one of these tests running?

File: change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json
Comment (atmgrifter00) [reply change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json:7]: To fix this I actually had to remove a call to `Math.floor`.

File: change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json
Comment (jattasNI) [reply change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json:1]: For the first issue, I can confirm that it's only reproducible in Storybook and not in the Angular example app. That may mean it's reproducible in cases where clients use iframes which is not ideal but also likely not a priority for us to fix. I found a couple threads that seem related but didn't look in detail whether there are fixes we could leverage:
- [SO post suggesting it's related to `preventDefault`](https://stackoverflow.com/questions/66703382/how-to-keep-watching-mouse-move-outside-iframe)
- [Firefox bug suggesting the behavior in other browsers is correct](https://bugzilla.mozilla.org/show_bug.cgi?id=1681647)

I'm fine to leave as-is if those threads don't yield a simple fix.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts:207]: 🤦‍♂️

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts]: The algorithm has changed to simply be:
currentFractionalWidth = (currentPixelWidth / initialPixelWidth) * initialFractionalWidth

So, now, if we have a series of fractional width columns: [1, 1, 1, 1], with an initial table total column width of 400 (and no minimum column sizes), and then size the first column by 150 pixels (it started off as 100), then the resulting fractional widths will be [2.5, .5, .5, .5].

This means that if I add a new column after the above sizing operation, it will have its expected default render width when considering its default fractional width of 1 (be twice as big as the smallest columns).

File: packages/nimble-components/src/table/testing/table.pageobject.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/testing/table.pageobject.ts:250]: Nothing on this line immediately jumps out to me as being possibly undefined. Can you double check that the `?? 0` is needed?

File: packages/nimble-components/src/table/testing/table.pageobject.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/testing/table.pageobject.ts:455]: Any reason not to simplify this condition?
```suggestion
        if (!divider) {
```

File: packages/nimble-components/src/table/testing/table.pageobject.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/testing/table.pageobject.ts:491]: Any reason not to simplify this condition?
```suggestion
        if (!divider) {
```

File: packages/nimble-components/src/table/testing/table.pageobject.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/testing/table.pageobject.ts:460]: `active` is a CSS class now rather than an attribute, right? Is this line still needed? If so, should there be an equivalent line setting `active` back to `false` at the end of the function?

Same question for `dragSizeColumnByLeftDivider` below.

File: packages/nimble-components/src/table/tests/table-column-sizing.spec.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/tests/table-column-sizing.spec.ts:373]: The test names and drag deltas here don't seem to match. For example, this test claims to be "sizing left" but provides a positive drag delta.

File: packages/nimble-components/src/table/tests/table-column-sizing.spec.ts
Comment (mollykreis) [inline packages/nimble-components/src/table/tests/table-column-sizing.spec.ts:371]: Is there a test that covers moving the mouse farther to the left than can columns can shrink, then moving the mouse to the right a little bit and ensuring column widths don't change (i.e. that the divider gets re-synced to the cursor's position)? I didn't see one, but it's quite possible I missed it. If there isn't one, could you add one?

File: packages/nimble-components/src/table/index.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/index.ts]: Nope, we don't.

File: packages/nimble-components/src/table/index.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/index.ts]: Nope. Removed.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts]: I've changed it to `requestedResizeAmount`, but I'm wondering if @jattasNI or @rajsite have other suggestions that might better indicate that this value can be negative. For whatever reason, I think a negative "amount" sounds odd, whereas a negative "delta" does not.

File: packages/nimble-components/src/table/testing/table.pageobject.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/testing/table.pageobject.ts]: Removed.

File: packages/nimble-components/src/table/tests/table-column-sizing.spec.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/tests/table-column-sizing.spec.ts:371]: Added test.

File: change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json
Comment (jattasNI) [inline change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json:1]: Reminder to update your PR description (it says tests are still in work)

File: change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json
Comment (jattasNI) [inline change/@ni-nimble-components-752d1b02-6122-4d2f-b737-809aff6729d2.json:2]: I updated from main again to see if it fixes the spurious Chromatic diffs.

**Review by jattasNI — APPROVED** [review]
Given that Molly's out and has given this a thorough review, I think it's ok to submit without her official approval (once Milan approves)

File: packages/nimble-components/src/table/template.ts
Comment (rajsite) [inline packages/nimble-components/src/table/template.ts:44]: Can we move more of the css logic into the stylesheet and keep the property definition to a minimum here?
For example:
```
--ni-private-table-scrollable-min-width: ${x => `${x.tableScrollableMinWidth}px)`}
```

And then we already have headerContianerMarginRight as a property `--ni-private-table-header-scrollbar-spacer-width`

style:
```
min-width: max(100%, calc(var(--ni-private-table-scrollable-min-width) + var(--ni-private-table-header-scrollbar-spacer-width));
```

I think that also illustrates as issue with the mental model here. These css custom property definitions are trying to define bindings to property values that can be used in styles, not styles themselves. They are just making observable properties usable in CSS.

So we may want to rename `--ni-private-table-header-scrollbar-spacer-width` to something that better aligns to the property it corresponds to, ie `--ni-private-table-header-container-margin-right`.
Then it'll be easier to say in the CSS that the `.header-row` `min-wdith` is a function of the right margin and scrollable min width, etc.

Wouldn't mind a pass in this PR or a separate one that changed the names of most of these to better align to the property they are exposing. I realize it doesn't work perfectly, for example for booleans and the row grid columns, but I think we can improve the correlation a bit.

File: packages/nimble-components/src/table/template.ts
Comment (rajsite) [inline packages/nimble-components/src/table/template.ts:76]: think you can specify the parent type here and avoid some `as` casts
```suggestion
                            ${repeat(x => x.visibleColumns, html<TableColumn, Table>`
```

File: packages/nimble-components/src/table/template.ts
Comment (rajsite) [inline packages/nimble-components/src/table/template.ts:85]: with the above can likely skip the `as` Table cast

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (rajsite) [inline packages/nimble-components/src/table/models/table-layout-manager.ts:80]: Extraneous context:

In the [WebVI days](https://dev.azure.com/ni/DevCentral/_git/ASW?path=%2FSource%2FVI%2FHtmlControls.Design%2FWeb%2FFramework%2FniUIActivityService.js&_a=contents&version=GBmain) found there are a bunch of extra cases to deal with. mouseup is not guaranteed so should also listen for window blur (and probably other events) and then there is the slew separate touch events which were deprecated. Which was funny because touch events were deprecated and the replacement api pointer events wasn't standard yet 🤣 BUT TODAY...

Recommendation:

We should switch to [pointer events](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events) which are mostly analogous to mouse events but provide a unified api across mouse, touch, arbitrary "pointer" concepts.

In particular this use-case seems well suited for the [setPointerCapture](https://developer.mozilla.org/en-US/docs/Web/API/Element/setPointerCapture) workflow which will intercept all other events for us, guarantee the life cycle completes, and actually might obviate the need for the pane of glass (if true, sorry it didn't occur to me sooner 😵)

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (rajsite) [inline packages/nimble-components/src/table/models/table-layout-manager.ts:81]: Have we thought of an intent to cancel the column resize? I found myself wanting to push the `esc` to abort an ongoing resize operation.

Comment (rajsite) [reply packages/nimble-components/src/table/models/table-layout-manager.ts:81]: Might be pretty straightforward to represent under pointer events. i.e. the esc key would just restore the sizes and releasePointerCapture (which fires pointercancel I believe) and leaves you ready for the next pointerdown.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts:80]: Agreed. I've filed [this issue](https://github.com/ni/nimble/issues/1362) for investigating the usage of pointer events.

File: packages/nimble-components/src/table/models/table-layout-manager.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/models/table-layout-manager.ts:81]: I think we can tackle this along with the pointer events I've created [this issue](https://github.com/ni/nimble/issues/1362) for.

**Review by rajsite — APPROVED** [review]

**Pull request merged**
